### PR TITLE
Implemented extensible Schnorr signer

### LINKED ISSUE #237: Schnorr signer crate
There are a couple projects requiring a rust mina signer: websnapps and chainsafe

In websnapps the TLS notary needs to sign notarizations. Ideally, these should use the standard mina schnorr signature algorithm.

Instead of using the c reference signer (which was never intended to be a library) and would require many changes, we should implement a library for signing transactions and notarizations.

Since Chainsafe's mina-rust implementation also needs a mina signature library it makes the most sense to do this in rust so that both projects can reuse the same implementation.

More details: https://github.com/MinaProtocol/mina/issues/9594

### PR DESCRIPTION
**Pull request by jspada**
* Mina's Schnorr signature scheme
    * Random oracle input serializer
    * Hashable and Signable interfaces
    * Standard signature unit tests

Closes #237

### CODE REVIEW

File: signer/README.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+This crate provides an API and framework for Mina signing.  It follows the algorithm outlined in the [Mina Signature Specification](https://github.com/MinaProtocol/mina/blob/develop/docs/specs/signatures/description.md).
+
+## Simple interface
+
+The simple interface uses the default signer configuration compatible with mainnet and testnet transaction signatures.
+
+```rust
+use rand;
+use mina_signer::{NetworkId, Keypair, Signer};
+use mina_signer::NetworkId;
```
Comment (mimoo) [inline signer/README.md:12]: importing NetworkId twice

File: signer/README.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+let mut ctx = mina_signer::create(NetworkId::TESTNET);
+let sig = ctx.sign(key_pair, transaction);
+
+assert_eq!(ctx.verify(sig, key_pair.public, transaction), true);
```
Comment (mimoo) [inline signer/README.md:17]: you can just write `assert!(ctx.verify(sig, key_pair.public, transaction))`

File: signer/README.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+The advanced interface allows specification of an alternative cryptographic sponge and parameters, for example, in order to create signatures that can be verified more efficiently using the Kimchi proof system.
+
+```rust
+use rand;
+use mina_signer::{NetworkId, Keypair, Signer};
+use oracle::{pasta, poseidon};
+
+let mut ctx = mina_signer::custom::<poseidon::PlonkSpongeConstants5W>(
+    pasta::fp5::params(),
```
Comment (mimoo) [inline signer/README.md:30]: why is poseidon stuff being passed twice, once as generic param and once as argument?

File: signer/README.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+impl Hashable for Foo {
+    fn to_roinput(self) -> ROInput {
+        let mut roi = ROInput::new();
+
+        roi.append_u32(self.foo);
+        roi.append_u64(self.bar);
+
+        roi
```
Comment (mimoo) [inline signer/README.md:60]: did you create an issue to use serde for that :o ? This manual redefinition of the struct really scares me as we change these structs all the time

File: signer/README.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    }
+}
+
+impl Signable for Foo {
+    fn domain_string(network_id: NetworkId) -> &'static str {
+       match network_id {
+           NetworkId::MAINNET => "FooSigMainnet",
+           NetworkId::TESTNET => "FooSigTestnet",
```
Comment (mimoo) [inline signer/README.md:68]: it feels a bit tedious to define a domain separation string for each network, why is this needed? Aren't we using the network id elsewhere to create different hash functions?

File: signer/src/domain.rs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+//!
+//! Shorthands and helpers for base and scalar field elements
+
+use ark_ec::AffineCurve;
+use ark_ff::PrimeField; // for into_repr()
+
+use mina_curves::pasta::pallas as Pallas;
+
+/// Affine curve point type
+pub use Pallas::Affine as CurvePoint;
```
Comment (mimoo) [inline signer/src/domain.rs:11]: why re-export this type?

File: signer/src/domain.rs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+/// Base field element type
+pub type BaseField = <CurvePoint as AffineCurve>::BaseField;
+/// Scalar field element type
+pub type ScalarField = <CurvePoint as AffineCurve>::ScalarField;
+
+use ark_serialize::CanonicalSerialize;
```
Comment (mimoo) [inline signer/src/domain.rs:17]: move all the `use` above :o

File: signer/src/domain.rs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+/// Field element helpers
+pub trait FieldHelpers<F: PrimeField> {
```
Comment (mimoo) [inline signer/src/domain.rs:20]: can you move this to o1-utils?

File: signer/src/signature.rs [hunk 9]
```diff
@@ -0,0 +1,33 @@
+//! Mina signature structure and associated helpers
+
+use std::fmt;
+
+use crate::{BaseField, FieldHelpers, ScalarField};
+
+/// Signature structure
+#[derive(Clone, Copy, Eq, fmt::Debug, PartialEq)]
+pub struct Signature {
+    /// Field component
```
Comment (mimoo) [inline signer/src/signature.rs:10]: you mean _base_ field component

File: signer/src/signature.rs [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    pub rx: BaseField,
+
+    /// Scalar component
+    pub s: ScalarField,
```
Comment (mimoo) [inline signer/src/signature.rs:14]: do these fields need to be public outside of the crate? If not you can set `pub(crate)`

File: signer/src/seckey.rs [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+use crate::ScalarField;
+
+/// Secret key
+#[derive(Clone, Copy, PartialEq, Eq)] // No Debug nor Display
+pub struct SecKey(ScalarField);
+
+impl SecKey {
+    /// Create a secret key from scalar field element
+    pub fn new(scalar: ScalarField) -> Self {
```
Comment (mimoo) [inline signer/src/seckey.rs:11]: no function to generate one randomly?

File: signer/src/seckey.rs [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+        Self(scalar)
+    }
+
+    /// Convert secret key into scalar field element
+    pub fn to_scalar(self) -> ScalarField {
```
Comment (mimoo) [inline signer/src/seckey.rs:16]: since this consumes self, this should be called `into_scalar` (rust convention)

File: signer/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,173 @@
+#![warn(missing_docs)]
+
+//! Mina signer library for verification and signing
```
Comment (mimoo) [inline signer/src/lib.rs:3]: btw, if you don't want to repeat yourself between the README and the doc here, you can also include some of the README content with this: https://github.com/o1-labs/proof-systems/blob/master/ocaml/ocaml-gen/src/lib.rs#L3

File: signer/src/keypair.rs [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+    pub fn rand(rng: &mut (impl RngCore + CryptoRng)) -> Self {
+        let secret: ScalarField = ScalarField::rand(rng);
+        let public: CurvePoint = CurvePoint::prime_subgroup_generator()
+            .mul(secret)
+            .into_affine();
+
+        Keypair {
+            secret: SecKey::new(secret),
+            public: PubKey::new(public),
+        }
```
Comment (mimoo) [inline signer/src/keypair.rs:40]: or you can call `Self::new(secret, public)`

File: signer/src/keypair.rs [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+use crate::{CurvePoint, FieldHelpers, PubKey, ScalarField, SecKey};
+use ark_ec::{AffineCurve, ProjectiveCurve};
+use ark_ff::UniformRand;
+use rand::{self, CryptoRng, RngCore};
+
+/// Keypair structure
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct Keypair {
+    /// Secret key
+    pub secret: SecKey,
```
Comment (mimoo) [inline signer/src/keypair.rs:16]: any reason this is `pub`?

File: README.md [hunk 16]
```diff
@@ -16,5 +16,6 @@ proof-systems/
 ├── groupmap/ # TODO: description
 ├── ocaml-gen/ # ocaml bindings generator tool
 ├── oracle/ # implementation of the poseidon hash function
+├── signer/ # implementation of schnorr signature scheme
```
Comment (mimoo) [inline README.md:19]: 👍

File: signer/src/pubkey.rs [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+use std::ops::Neg;
+
+use crate::{BaseField, CurvePoint, FieldHelpers};
+
+/// Length of Mina addresses
+pub const MINA_ADDRESS_LEN: usize = 55;
+
+/// Public key
+#[derive(Copy, Clone, fmt::Debug, PartialEq, Eq)]
+pub struct PubKey(CurvePoint);
```
Comment (mimoo) [inline signer/src/pubkey.rs:18]: 👍

File: signer/src/pubkey.rs [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+use sha2::{Digest, Sha256};
+use std::ops::Neg;
+
+use crate::{BaseField, CurvePoint, FieldHelpers};
+
+/// Length of Mina addresses
+pub const MINA_ADDRESS_LEN: usize = 55;
+
+/// Public key
+#[derive(Copy, Clone, fmt::Debug, PartialEq, Eq)]
```
Comment (mimoo) [inline signer/src/pubkey.rs:17]: `Debug` is already in scope (no need for `fmt::`)

File: signer/src/pubkey.rs [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+pub struct PubKey(CurvePoint);
+
+impl PubKey {
+    /// Create a public key from curve point
+    pub fn new(point: CurvePoint) -> Self {
```
Comment (mimoo) [inline signer/src/pubkey.rs:22]: is it important to do any validation check at this point? what if that pubkey is untrusted?

File: signer/src/pubkey.rs [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+    pub fn from_address(address: &str) -> Result<Self, &'static str> {
+        if address.len() != MINA_ADDRESS_LEN {
+            return Err("Invalid address length");
+        }
+
+        let bytes = bs58::decode(address)
+            .into_vec()
+            .map_err(|_| "Invalid address encoding")?;
+
+        let (raw, checksum) = (&bytes[..bytes.len() - 4], &bytes[bytes.len() - 4..]);
```
Comment (mimoo) [inline signer/src/pubkey.rs:36]: if bs58 is variable-length encoding, don't you want to check the length of `bytes` instead of `address`? As bytes might be different length eventhough address is MINA_ADDRESS_LEN?

File: signer/src/pubkey.rs (hunk 19 above)
Comment (mimoo) [reply signer/src/pubkey.rs:22]: also it feels like this function should be renamed `from` or something with `from_` as this is external input

File: signer/src/pubkey.rs [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+            return Err("Invalid address version info");
+        }
+
+        let x = BaseField::from_bytes(x_bytes).map_err(|_| "invalid x-coordinate bytes")?;
+        let mut pt =
+            CurvePoint::get_point_from_x(x, y_parity).ok_or("Invalid address x-coordinate")?;
+
+        if pt.y.into_repr().is_even() == y_parity {
+            pt.y = pt.y.neg();
+        }
```
Comment (mimoo) [inline signer/src/pubkey.rs:57]: this public point verification logic should live in Pubkey::new

File: signer/src/pubkey.rs [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+
+        Ok(PubKey::new(pt))
+    }
+
+    /// Convert public key into curve point
+    pub fn to_point(self) -> CurvePoint {
```
Comment (mimoo) [inline signer/src/pubkey.rs:63]: `into_point`

File: signer/src/pubkey.rs [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+        self.0
+    }
+
+    /// Convert public key into compressed public key
+    pub fn to_compressed(self) -> CompressedPubKey {
```
Comment (mimoo) [inline signer/src/pubkey.rs:68]: `into_compressed`

File: signer/src/pubkey.rs [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+        let point = self.to_point();
+        CompressedPubKey {
+            x: point.x,
+            is_odd: !point.y.into_repr().is_even(),
```
Comment (mimoo) [inline signer/src/pubkey.rs:72]: why not use `is_odd`?

File: signer/src/pubkey.rs [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+        }
+    }
+
+    /// Serialize public key into corresponding Mina address
+    pub fn to_address(self) -> String {
```
Comment (mimoo) [inline signer/src/pubkey.rs:77]: into_adress

File: signer/src/pubkey.rs [hunk 26]
```diff
@@ -0,0 +1,152 @@
...
+#[derive(Clone, Copy)]
+pub struct CompressedPubKey {
+    /// X-coordinate
+    pub x: BaseField,
+
+    /// Parity of y-coordinate
+    pub is_odd: bool,
+}
+
+fn to_address(x: BaseField, is_odd: bool) -> String {
```
Comment (mimoo) [inline signer/src/pubkey.rs:105]: is there a rationale for implementing this outside of the `impl PublicKey`?

File: signer/src/roinput.rs [hunk 27]
```diff
@@ -0,0 +1,867 @@
...
+///         roi
+///     }
+/// }
+/// ```
+/// **Details:** For technical reasons related to our proof system and performance, fields are
+/// serialized for signing differently than other types. Additionally, during signing all members
+/// of the random oracle input get serialized together in two different ways: both as *bytes* and
+/// as a vector of *field elements*.  The random oracle input encapsulates and automates this
+/// complexity.
+
```
Comment (mimoo) [inline signer/src/roinput.rs:47]: remove space

File: signer/src/roinput.rs [hunk 28]
```diff
@@ -0,0 +1,867 @@
...
+            acc
+        });
+
+        bits.extend(&self.bits);
+
+        bits.into()
+    }
+
+    /// Serialize random oracle input to vector of base field elements
+    pub fn to_fields(&self) -> Vec<BaseField> {
```
Comment (mimoo) [inline signer/src/roinput.rs:110]: I think I pointed that before, but this will have collisions if you don't include the length of bits somewhere

File: signer/src/schnorr.rs [hunk 29]
```diff
@@ -0,0 +1,156 @@
...
+        let mut bytes = format!("{:*<20}", domain_string).as_bytes().to_vec();
+        bytes.resize(32, 0);
+
+        bytes
+    }
+
+    // This function uses a cryptographic hash function to create a uniformly and
+    // randomly distributed nonce.  It is crucial for security that no two different
+    // messages share the same nonce.
+    fn blinding_hash<H>(&self, kp: &Keypair, input: H) -> ScalarField
```
Comment (mimoo) [inline signer/src/schnorr.rs:98]: can you call this something else? like `derive_nonce` or something :D?

File: signer/src/schnorr.rs [hunk 30]
```diff
@@ -0,0 +1,156 @@
...
+        bytes[bytes.len() - 1] &= 0b0011_1111;
+
+        ScalarField::from_random_bytes(&bytes[..]).expect("failed to create scalar from bytes")
+    }
+
+    // This function uses a cryptographic hash function (based on a sponge construction) to
+    // convert the message to be signed (and some other information) into a uniformly and
+    // randomly distributed scalar field element.  It uses Mina's variant of the Poseidon
+    // SNARK-friendly cryptographic hash function.
+    // Details: <https://github.com/o1-labs/cryptography-rfcs/blob/httpsnapps-notary-signatures/mina/001-poseidon-sponge.md>
```
Comment (mimoo) [inline signer/src/schnorr.rs:127]: you should still use `///` btw, even on internal functions (as we can export doc for that too)

File: signer/tests/notarization.rs [hunk 31]
```diff
@@ -0,0 +1 @@
+
```
Comment (mimoo) [inline signer/tests/notarization.rs:1]: delete this file

File: signer/src/schnorr.rs [hunk 32]
```diff
@@ -0,0 +1,156 @@
...
+
+use crate::{
+    BaseField, CurvePoint, FieldHelpers, Hashable, Keypair, NetworkId, PubKey, ROInput,
+    ScalarField, Signable, Signature, Signer,
+};
+
+/// Schnorr signer context for the Mina signature algorithm
+///
+/// For details about the signature algorithm please see [crate::schnorr]
+pub struct Schnorr<SC: SpongeConstants> {
```
Comment (mimoo) [inline signer/src/schnorr.rs:35]: `struct SignerCtx` perhaps?

File: signer/src/schnorr.rs [hunk 33]
```diff
@@ -0,0 +1,156 @@
...
+        roi.append_bytes(&[self.network_id.into()]);
+
+        hasher.update(roi.to_bytes());
+
+        let mut bytes = [0; 32];
+        hasher.finalize_variable(|out| bytes.copy_from_slice(out));
+        // Drop the top two bits to convert into a scalar field element
+        //   N.B. Since the order p is very close to 2^m for some m, truncating only creates
+        //   a tiny amount of bias that should be insignificant and keeps the implementation
+        //   simple by avoiding reduction modulo p.
```
Comment (mimoo) [inline signer/src/schnorr.rs:117]: two other things you can point out:

- that this is true for Pallas' scalar field at least (not necessarily true if we change curve)
- that doing a modular reduction would be worse

File: signer/src/schnorr.rs [hunk 34]
```diff
@@ -0,0 +1,156 @@
...
+        S: Signable,
+    {
+        let mut roi: ROInput = input.to_roinput();
+        roi.append_field(pub_key.to_point().x);
+        roi.append_field(pub_key.to_point().y);
+        roi.append_field(rx);
+
+        // Set sponge initial state (explicitly init state so signer context can be reused)
+        // N.B. Mina sets the sponge's initial state by hashing the input's domain bytes
+        self.sponge.state = vec![BaseField::zero(); self.sponge.state.len()];
```
Comment (mimoo) [inline signer/src/schnorr.rs:139]: can you implement a `reset` function on `ArithmeticSponge` instead?

File: signer/src/schnorr.rs [hunk 35]
```diff
@@ -0,0 +1,156 @@
...
+        self.sponge.sponge_state = SpongeState::Absorbed(0);
+        self.sponge
+            .absorb(&[
+                BaseField::from_bytes(&Schnorr::<SC>::domain_bytes::<S>(self.network_id))
+                    .expect("invalid domain bytes"),
+            ]);
+        self.sponge.squeeze();
+
+        // Absorb random oracle input
+        self.sponge.absorb(&roi.to_fields());
```
Comment (mimoo) [inline signer/src/schnorr.rs:149]: do you think it'd make sence to have `to_fields()` consume `roi` so you can only call it once?

Comment (mimoo) [reply signer/src/schnorr.rs:149]: I really don't like the ROI approach btw because it feels prone to misuse, it looks like we're just shoving bits and field elements together and there might be collisions or ambiguity in the way we do this.

I think a better design would do this:

* absorb a field just absorbs a field element
* if you absorb bits/bytes, it caches them in a queue (and also prepends the length of the bits/bytes every time you call these functions in the queue as well)
* if you absorb bits/bytes when the queue gets bigger than the size of a field element, you convert some of the queue to a field element
* if you absorb field (or call `to_fields()`) when the bits queue is not empty, then you first convert your bit queue into field, stick that into the field array, and then absorb your field

**Review by mimoo — APPROVED** [review]
good job on that implementation! Left a bunch of comments :)

File: signer/README.md (hunk 3 above)
Comment (jspada) [reply signer/README.md:30]: This is just from how proof-systems sponge works.  They are two different things.  The generic param is the `SpongeConstants`, which configures the sponge capacity, rate, rounds, s-box alpha, width and a couple of other things.  The argument is the `ArithmeticSpongeParams` which just contains the round constants and MDS matrix.  If the sponge, it's consts and params are ever refactored into a single parameter then we can change this API to just use the single thing.  For now, unfortunately they come as two things from two different places.

File: signer/README.md (hunk 5 above)
Comment (jspada) [reply signer/README.md:68]: Unfortunately there's no network id in the signed data, which is why mina does this.  There are some interesting historic reasons for this.  We discussed improving this previously, but it would entail a bunch of 3rd parties having to update their implementations (perhaps some are even on custom hardware), which was not ideal.

File: signer/README.md (hunk 4 above)
Comment (jspada) [reply signer/README.md:60]: Yes.  https://github.com/o1-labs/proof-systems/issues/241

Agreed!

For now this should be OK because for tx signatures it doesn't change.  For backward compatibility, each version of tx/signatures is a new thing (i.e. new tx version and structure), rather than updating the old one.

Anyway, it's way better to do it automatically and I will try to do this with subsequent ticket 241 :-)

File: signer/src/domain.rs (hunk 8 above)
Comment (jspada) [reply signer/src/domain.rs:20]: Yes, I plan to do this in this ticket https://github.com/o1-labs/proof-systems/issues/242

File: signer/src/signature.rs (hunk 9 above)
Comment (jspada) [reply signer/src/signature.rs:10]: Y

File: signer/src/signature.rs (hunk 10 above)
Comment (jspada) [reply signer/src/signature.rs:14]: Some things in another crate (the tests and the websnapps prover) need to access them individually.   Seems like something that should be public.

File: signer/src/seckey.rs (hunk 11 above)
Comment (jspada) [reply signer/src/seckey.rs:11]: Wish granted :-)

File: signer/src/lib.rs [hunk 36]
```diff
@@ -0,0 +1,173 @@
+#![warn(missing_docs)]
```
[the remark this answers is not in the document]
Comment (jspada) [reply signer/src/lib.rs:1]: 👍🏼

File: signer/src/lib.rs (hunk 13 above)
Comment (jspada) [reply signer/src/lib.rs:3]: rust is awesome

File: signer/src/pubkey.rs [hunk 37]
```diff
@@ -0,0 +1,152 @@
+//! Public key structures and algorithms
+//!
+//! Definition of public key structure and helpers
+
+use ark_ff::{BigInteger, PrimeField};
+use bs58;
```
[the remark this answers is not in the document]
Comment (jspada) [reply signer/src/pubkey.rs:6]: haha!

File: signer/src/keypair.rs (hunk 15 above)
Comment (jspada) [reply signer/src/keypair.rs:16]: Code in schnorr.rs needs access to secret.

File: signer/README.md (hunk 5 above)
Comment (mimoo) [reply signer/README.md:68]: I think this should be documented in the trait, to make sure people choose different domain string depending on the network

File: signer/src/keypair.rs (hunk 15 above)
Comment (jspada) [reply signer/src/keypair.rs:16]: I'll `pub(crate)` it

Comment (mimoo) [reply signer/src/keypair.rs:16]: Ideally this should not be public IMO. Two things you could do:

* make it public only to the crate with `pub(crate)`, but you can still write code that will extract the secret key within the crate
* make it public only to this file, by removing `pub` and implementing signing on `Keypair`.

File: signer/src/pubkey.rs (hunk 20 above)
Comment (jspada) [reply signer/src/pubkey.rs:36]: AFAIK MINA_ADDRESS_LEN is on the b58 encoding, but I can add an extra check for bytes.

File: signer/src/pubkey.rs (hunk 26 above)
Comment (jspada) [reply signer/src/pubkey.rs:105]: So it can be shared by `PubKey` and `CompressedPubKey`.

File: signer/src/roinput.rs (hunk 28 above)
Comment (jspada) [reply signer/src/roinput.rs:110]: Yes, but I can't change this or it will not be compatible with existing mainnet/testnet signatures.  If we want to fix this we should introduce tx version 2 and definitely pad with the length.

File: signer/src/schnorr.rs (hunk 29 above)
Comment (jspada) [reply signer/src/schnorr.rs:98]: how about `nonce_hash` ?

File: signer/src/schnorr.rs (hunk 32 above)
Comment (jspada) [reply signer/src/schnorr.rs:35]: Schnorr implements Signer interface, so it can be called whatever it wants as long as it implements the Signer interface.  I thought this is nice, so we can at this level just call things what they are under the hood, which isn't really visible to the user of the API

File: signer/src/schnorr.rs (hunk 34 above)
Comment (jspada) [reply signer/src/schnorr.rs:139]: Agreed - can this be part of this PR?

File: signer/src/schnorr.rs (hunk 33 above)
Comment (jspada) [reply signer/src/schnorr.rs:117]: 👍🏼

File: signer/src/schnorr.rs (hunk 35 above)
Comment (jspada) [reply signer/src/schnorr.rs:149]: I agree with your suggestions, but I think this should be done in another ticket.  The reason is because if we cannot make changes to this without breaking compatibility with mainnet mina transaction signatures, which is one of my main objectives (to be the same as the OCaml code).  That said, these are very good suggestions -- especially the length-padding part.  Thus, I've opened up a separate ticket for improving and simplifying the ROI structure and length padding things.  https://github.com/o1-labs/proof-systems/issues/244

File: signer/src/pubkey.rs (hunk 19 above)
Comment (jspada) [reply signer/src/pubkey.rs:22]: Good one!  Agreed -- thanks.

**Review by mrmr1993 — APPROVED** [review]

File: oracle/src/poseidon.rs [hunk 38]
```diff
@@ -84,6 +84,7 @@ pub trait Sponge<Input: Field, Digest> {
     fn new(params: ArithmeticSpongeParams<Input>) -> Self;
     fn absorb(&mut self, x: &[Input]);
     fn squeeze(&mut self) -> Digest;
+    fn reset(&mut self);
```
Comment (mimoo) [inline oracle/src/poseidon.rs:87]: missing some doc ;)
